### PR TITLE
Make Array contract application lazy

### PR DESCRIPTION
**Pull request by fuzzypixelz**
# The problem

Right now applying an Array contract does something that resembles:

```
%map% t (fun value => %assume% elt (%go_array% l) value)
```

This essentially puts the array elements behind a mapping of the contract on each element. Meaning that simply taking the head of the array will trigger `n = array.length xs` operations, even though we only need one (this is because `map` is a strict primitive operation implemented in the interpreter). Also, if we take the head of the array multiple times, we will needlessly apply the contract again and again.

# How this improves the situation

When an Array receives a contract application, the contract is simply stored in its attributes and only applied when data leaves the Array, or when we merge with another Array with potentially incompatible contracts.

# Issues

This code seems to break `DeepSeq`, especially when serializing terms. This is addressed in https://github.com/tweag/nickel/pull/767 by adding a proper `Force` operator.

### CODE REVIEW

**Comment by fuzzypixelz** [comment]
Sure. I'll close the original PR.

File: src/term.rs [hunk 1]
```diff
@@ -183,14 +183,69 @@ impl Default for BindingType {
...
+        self
+    }
+
+    /// Drop the pending contracts.
+    pub fn contracts_cleared(mut self) -> Self {
+        self.pending_contracts.clear();
+        self
+    }
+
+    /// Extend contracts from an iterator of `PendingContract`.
```
Comment (yannham) [inline src/term.rs:234]:
```suggestion
    /// Extend contracts from an iterator of `PendingContract`.
    /// De-duplicate equal contracts. Note that current contract
    /// equality testing is very limited, but this may change in the
    /// future
```

File: src/term.rs [hunk 2]
```diff
@@ -836,17 +891,34 @@ pub enum UnaryOp {
     /// Transform a string to an enum.
     EnumFromStr(),
     /// Test if a regex matches a string.
-    /// Like [`StrMatch`], this is a unary operator because we would like a way to share the
+    /// Like [`UnaryOp::StrMatch`], this is a unary operator because we would like a way to share the
```
Comment (yannham) [inline src/term.rs:894]: Just out of curiosity, didn't `` [`StrMatch`] `` work? I thought we tested those links when building the documentation of 0.1.0.

File: src/term.rs [hunk 3]
```diff
@@ -836,17 +891,34 @@ pub enum UnaryOp {
...
     /// Match a regex on a string, and returns the captured groups together, the index of the
     /// match, etc.
     StrMatch(),
-    /// Version of `StrIsMatch` which remembers the compiled regex.
+    /// Version of [`UnaryOp::StrIsMatch`] which remembers the compiled regex.
     StrIsMatchCompiled(CompiledRegex),
-    /// Version of `StrMatch` which remembers the compiled regex.
+    /// Version of [`UnaryOp::StrMatch`] which remembers the compiled regex.
     StrMatchCompiled(CompiledRegex),
+    /// Force "full" evaluation of a term and return it.
```
Comment (yannham) [inline src/term.rs:905]:
```suggestion
    /// Force full evaluation of a term and return it.
```

File: src/term.rs [hunk 4]
```diff
@@ -836,17 +891,34 @@ pub enum UnaryOp {
...
+    ///
+    /// This was added in the context of [`BinaryOp::ArrayLazyAssume`],
+    /// and may not make much sense on its own.
```
Comment (yannham) [inline src/term.rs:908]:
```suggestion
    /// This was added in the context of [`BinaryOp::ArrayLazyAssume`],
    /// in particular to make serialization work with lazy array contracts.
```

I'm not sure it doesn't make sense. It may make more sense than the current `deep_seq`, somehow (it evaluates more). If I may make a suggestion, we could stick to the motivation :slightly_smiling_face:

File: src/typecheck/operation.rs [hunk 5]
```diff
@@ -195,6 +195,8 @@ pub fn get_uop_type(
                 ("groups", mk_typewrapper::array(AbsType::Str()))
             ),
         ),
+        // Dyn -> Dyn
+        UnaryOp::Force(_) => (mk_typewrapper::dynamic(), mk_typewrapper::dynamic()),
```
Comment (yannham) [inline src/typecheck/operation.rs:199]: Why does this appear in this PR?

File: src/eval/operation.rs [hunk 6]
```diff
@@ -1851,15 +1964,53 @@ fn process_binary_operation(
...
+                                apply_contracts(t, ctrs_left.clone(), pos1)
+                                .closurize(&mut env, env1.clone())
+                            ));
+
+                            ts.extend(ts2.into_iter().map(|t|
+                                apply_contracts(t, ctrs_right.clone(), pos2)
+                                .closurize(&mut env, env2.clone())
+                            ));
+
+                            let ctrs_common = attrs1
```
Comment (yannham) [inline src/eval/operation.rs:2001]: I think you can use `std::iter::partition` above to compute `ctrs_left` and `ctrs_common` at once above, instead of doing another pass here. Maybe that will saved one `cloned()` as well

File: src/eval/operation.rs [hunk 7]
```diff
@@ -2145,6 +2301,58 @@ fn process_binary_operation(
...
+                        term: t2,
+                        pos: pos2,
+                    },
+                ))
+            };
+
+            match_sharedterm! {t2,
+                with {
+                    Term::Array(ts, attrs) => {
+
```
Comment (yannham) [inline src/eval/operation.rs:2331]:
```suggestion
```

File: src/term.rs [hunk 8]
```diff
@@ -183,14 +183,69 @@ impl Default for BindingType {
...
+    }
+
+    /// Drop the pending contracts.
+    pub fn contracts_cleared(mut self) -> Self {
+        self.pending_contracts.clear();
+        self
+    }
+
+    /// Extend contracts from an iterator of `PendingContract`.
+    pub fn with_contracts<I>(mut self, iter: I) -> Self
```
Comment (yannham) [inline src/term.rs:235]: I find the name to be slightly ambiguous. Usually `with_xxx` means that `xxx` will be set to the new parameter (a form of `set` + return self), while here it _extends_ the current one. Maybe a name like `with_extra_contracts` or the like would be clearer?

File: src/eval/operation.rs [hunk 9]
```diff
@@ -2408,11 +2616,55 @@ fn eq(env: &mut Environment, c1: Closure, c2: Closure) -> EqResult {
...
         }
-        (Term::Array(l1, _), Term::Array(l2, _)) if l1.len() == l2.len() => {
+        (Term::Array(l1, a1), Term::Array(l2, a2)) if l1.len() == l2.len() => {
             // Equalities are tested in reverse order, but that shouldn't matter. If it
             // does, just do `eqs.rev()`
-            let eqs = l1.into_iter().zip(l2.into_iter());
-            gen_eqs(eqs, env, env1, env2)
+
+            // We should apply all contracts here, otheriwse we risk having wrong values,
+            // think record contrats with default contracts, Wrapped terms, etc.
```
Comment (yannham) [inline src/eval/operation.rs:2624]:
```suggestion
            // think record contrats with default contracts, wrapped terms, etc.
```

**Review by yannham — APPROVED** [review]

File: src/term.rs (hunk 2 above)
Comment (fuzzypixelz) [reply src/term.rs:895]: No, they're only highlighted as code. I discovered this while adding documentation for `Force` and testing how it renders.

Just doubled-checked on master.

File: src/term.rs (hunk 8 above)
Comment (fuzzypixelz) [reply src/term.rs:238]: Oh I see! Is there some definitive reference for Rust API naming conventions?

File: src/typecheck/operation.rs (hunk 5 above)
Comment (fuzzypixelz) [reply src/typecheck/operation.rs:199]: Because we merged in `feature/force-operator`?

File: src/pretty.rs [hunk 10]
```diff
@@ -238,7 +238,7 @@ where
                         }, // be sure we have at least 1 `%` sign when an interpolation is present
                     )
                     .max()
-                    .unwrap();
+                    .unwrap_or(1);
```
Comment (yannham) [inline src/pretty.rs:241]: Can we get rid of the `else { 1 }` above if we use `1` as a default value here? But I think those kind of improvements, albeit small, would be better off as a separate PR.

File: src/term.rs [hunk 11]
```diff
@@ -183,14 +183,70 @@ impl Default for BindingType {
     }
 }
 
+/// A contract with its associated data.
+#[derive(Debug, PartialEq, Clone)]
+pub struct PendingContract {
+    /// The "pending" contract, can be a function or a record.
```
Comment (yannham) [inline src/term.rs:189]:
```suggestion
    /// The pending contract, can be a function or a record.
```

File: src/interner.rs [hunk 12]
```diff
@@ -0,0 +1,10 @@
+//! A string interner.
```
Comment (yannham) [inline src/interner.rs:1]: I guess this is unintended :slightly_smiling_face:

**Review by yannham — APPROVED** [review]

File: src/interner.rs (hunk 12 above)
Comment (fuzzypixelz) [reply src/interner.rs]: For sure!
